File: python3/minimum_time_to_make_rope_colorful.py
"""
Space   : O(1)
Time    : O(n)
"""


class Solution:
    def minCost(self, colors: str, neededTime: List[int]) -> int:
        ans = 0
        n = len(colors)

        cur = colors[0]
        sum_time = neededTime[0]
        max_time = neededTime[0]
        for i in range(1, n):
            if colors[i] == cur:
                sum_time += neededTime[i]
                max_time = max(max_time, neededTime[i])
            else:
                # resolve prev color
                if sum_time > max_time:
                    ans += sum_time - max_time
                cur = colors[i]
                sum_time = neededTime[i]
                max_time = neededTime[i]

        if sum_time > max_time:
            ans += sum_time - max_time

        return ans

File: golang/minimum_time_to_make_rope_colorful.go
package main

/*
Space	: O(1)
Time	: O(n)
*/

func minCost(colors string, neededTime []int) int {
	n := len(colors)
	ans := 0

	cur := colors[0]
	maxTime := neededTime[0]
	sumTime := neededTime[0]

	for i := 1; i < n; i++ {
		if colors[i] == cur {
			sumTime += neededTime[i]
			maxTime = Max(maxTime, neededTime[i])
		} else {
			if sumTime > maxTime {
				ans += sumTime - maxTime
			}
			cur = colors[i]
			sumTime = neededTime[i]
			maxTime = neededTime[i]
		}
	}

	if sumTime > maxTime {
		ans += sumTime - maxTime
	}

	return ans
}
